smartsim/_core/launcher/dragon/dragonBackend.py

```diff
@@ -63,10 +83,72 @@ class DragonBackend:
     by threads spawned by it.
     """
 
-    def __init__(self) -> None:
-        self._proc_groups: t.Dict[str, t.Tuple[ProcessGroup, t.List[int]]] = {}
+    def __init__(self, pid: int) -> None:
+        self._pid = pid
+        self._group_infos: t.Dict[str, ProcessGroupInfo] = {}
         self._step_id_lock = RLock()
+        self._hostlist_lock = RLock()
         self._step_id = 0
+        # hosts available for execution
+        # dictionary maps hostname to step_id of
+        # step being executed on it
+        self._initialize_hosts()
+        # Suppress pylint error which only happens for Python 3.7
+        # pylint: disable-next=unsubscriptable-object
+        self._queued_steps: collections.OrderedDict[str, DragonRunRequest] = (
+            collections.OrderedDict()
+        )
+        self._running_steps: t.List[str] = []
+        self._completed_steps: t.List[str] = []
+
+        num_hosts = len(self._hosts)
+        host_string = str(num_hosts) + (" hosts" if num_hosts > 1 else " host")
+        self._shutdown_requested = False
+        self._updates = 0
+        print(host_string + f" available for execution: {self._hosts}")
+
+    def print_status(self) -> None:
+        print(f"{self._updates}: System hosts: ", self._hosts)
+        print(f"{self._updates}: Free hosts: ", self._free_hosts)
+        print(f"{self._updates}: Allocated hosts: ", self._allocated_hosts)
+        print(f"{self._updates}: Running steps: ", self._running_steps)
+        print(f"{self._updates}: Group infos: ", self._group_infos)
+
+    @property
+    def should_shutdown(self) -> bool:
+        return self._shutdown_requested
+
+    def _initialize_hosts(self) -> None:
+        with self._hostlist_lock:
+            self._hosts: t.List[str] = sorted(
+                Node(node).hostname for node in System().nodes
+            )
+            self._free_hosts: t.List[str] = self._hosts.copy()
+            self._allocated_hosts: t.Dict[str, str] = {}
+
+    def _request_is_satisfiable(self, request: DragonRunRequest) -> bool:
+        """Check if request can be satisfied with resources
+        available in the allocation.
+        Currently only checks for total number of nodes,
+        in the future it will also look at other constraints
+        such as memory, accelerators, and so on.
+        """
+        return request.nodes <= len(self._hosts)
+
+    def _allocate_step(
+        self, step_id: str, request: DragonRunRequest
+    ) -> t.Optional[t.List[str]]:
+
+        num_hosts: int = request.nodes
+        with self._hostlist_lock:
+            if num_hosts > len(self._free_hosts):
```

Comment:
I know this condition should never be met due to the pydantic schema validation, but just to satisfy my paranoia, can we add something like this?
```suggestion
            if num_hosts <= 0 or num_hosts > len(self._free_hosts):
```

Reply:
Applied